### ISSUE TITLE
Allow mods to control which stats appear in UISoldierHeader

### ISSUE THREAD

**Issue by pledbrook** [X2CommunityCore/X2WOTCCommunityHighlander#533]
`UISoldierHeader` has too little to space to display all unit stats even in the base game (Where Art Thou, Defence?) Throw in Psi Offense and any others that might be added/supported by mods, it makes sense to allow mods to control which stats are displayed in the armoury header.

I suggest we create an event that sends an array of stat types (from enum `ECharStatType`), which can be reordered or have elements replaced by mods.

Another option would be to send an array of stat type & value pairs, so that mods can override the displayed value as well.

**Comment by pledbrook** [comment]
I wonder if this is something that @robojumper's Squad Select would also want to use with Show Me The Stats.

**Comment by Xymanek** [comment]
Instead of allowing mods to change the value, perhaps pass a bool that controls whether to show the max or not?

**Comment by pledbrook** [comment]
Isn't that what happens now? It displays max +/- modifier? Or is that just a LW thing?

**Comment by Xymanek** [comment]
The will is `current`/`max`

**Comment by pledbrook** [comment]
I don't have a strong opinion either way. Happy to add a bool. Although it would help to know how best to pass arrays of pairs via the tuple. Any suggestions?

**Comment by Xymanek** [comment]
> Although it would help to know how best to pass arrays of pairs via the tuple

You would need to pass tuple as `Tuple.Data[0].o`, `Tuple.Data[1].o`, etc. I would argue for a specific event data class, instead of a tuple

**Comment by jmartinez989** [comment]
ok so what I am thinking is to make an object that has 3 different items in it, one of which is the label for the stat and the other the value of the stat and the last is a bool which would determine whether or not we should show that stat. Then an array of those can be passed to SetSoldierStats (obviously the parameter list of the function would have to change).

**Comment by jmartinez989** [comment]
So I was writing some code to implement this functionality where I was going to fire an event in `UISoldierHeader.SetSoldierStats()` but then I found that another class called `UIInventory_Implants` has a function called `SelectedItemChanged()` which calls `SetSoldierStats()` as well which could cause a problem to how this would work.

this might not actually be too much of a problem though because we can always get a referrence to the unit that we will be getting stats in the `SetSoldierStats()` function and pass it to the event to order the stats. @Xymanek any input on this? Would you be ok with just getting the unit referrence in the `SetSoldierStats()` function of `UISoldierHeader`?

Also would we want an option to color the stat labels and values like how will and psi are?

**Comment by Xymanek** [comment]
The issue is that `SetSoldierStats` expects prepared list of stats, so triggering an event from there doesn't really work. I have a feeling that we will need to extract the logic for calculating the values of stats to a separate function and trigger the event from there

**Comment by pledbrook** [comment]
`SetSoldierStats()` would need to change regardless if we want to be able to display any stat not already supported by its argument list. I know that original LW2 modified its argument list to include Defence.

**Comment by Xymanek** [comment]
I think we should leave `SetSoldierStats()` alone and make a new, improved function which would dispatch the event and calculate the values. Additionally, I think that function should take an optional delegate to replace the logic for converting the stat to displayed value to accommodate `UIInventory_Implants`

**Comment by jmartinez989** [comment]
So do you mean that what should be done is to make a new function that will replace `SetSoldierStats()` and include `UIInventory_Implants` in the CHL so that we can change its call to `SetSoldierStats()` to the new function?

**Comment by Xymanek** [comment]
> make a new function that will replace `SetSoldierStats()`

Correct, although ~replace~ supercede

**Comment by jmartinez989** [comment]
ok so I will take all of the logic in `PopulateData()` which is the function that does all of the calculation of stats and move it to the new function along with triggering the event from there and then also displaying stats from that same function?

**Comment by pledbrook** [comment]
@Xymanek Why not a `SetSoldierStats_CH()` function that supports all the available stats? Introducing a delegate and a whole new function that works differently seems like a significant refactoring with little gain that I can see. In this case, I favour conservatism unless there is a strong argument for a more significant change.

**Comment by Xymanek** [comment]
The argument is to allow mods to change things like colours, show/hide max, etc. If we make a function that only accepts more stats (but in already formatted function) we lose that possibility. Furthermore, if we ever decide to make new stats (like with inventory slots), we would need to make `SetSoldierStats_CH_Improved()` (since we can't change signature of a public function).

That said, I can't deny that `SetSoldierStats_CH()` would def be a simpler solution

**Comment by pledbrook** [comment]
I understand now. `SetSoldierStats()` takes the rendered text. I can't say I'm thrilled at any old bit of code being able to change what is rendered. Recipe for inconsistency.

What if we centralised rendering in `UISoldierHeader` and allowed other code to simply provide alternative bonuses/penalties?

**Note** The code in `UIInventory_Implants` that calls `SetSoldierStats()` doesn't actually do anything. The `UISoldierHeader` isn't even visible while one hovers over the items in the PCS list. And I'm guessing the `UIArmory_Loadout` screen uses some other approach to display the different bonuses as you hover over different pieces of equipment.

In other words, I'm not sure we should be considering `UIInventory_Implants` at all when trying to work out how to make this improvement.

**Comment by pledbrook** [comment]
For reference, `UIArmory_Loadout` simply calls `UISoldierHeader.PopulateData()` with references to the item that the player is hovering over and the item it would replace if selected. I'm guessing this is what the PCS code should do too if `UISoldierHeader` were actually visible at the relevant time.
